src/server/replica.cc

```diff
@@ -550,40 +581,110 @@ error_code Replica::StartAsDflyFlow() {
   if (flow_directive == "FULL") {
     eof_token = ToSV(resp_args_[1].GetBuf());
   } else {
-    LOG(ERROR) << "Bad FLOW response " << ToSV(io_buf->InputBuffer());
+    LOG(ERROR) << "Bad FLOW response " << ToSV(leftover_buf_->InputBuffer());
   }
-  io_buf->ConsumeInput(consumed);
+  leftover_buf_->ConsumeInput(consumed);
 
   state_mask_ = R_ENABLED | R_TCP_CONNECTED;
 
   // We can not discard io_buf because it may contain data
   // besides the response we parsed. Therefore we pass it further to ReplicateDFFb.
-  sync_fb_ =
-      ::boost::fibers::fiber(&Replica::FullSyncDflyFb, this, std::move(io_buf), move(eof_token));
+  sync_fb_ = ::boost::fibers::fiber(&Replica::FullSyncDflyFb, this, sb, move(eof_token));
 
   return error_code{};
 }
 
-void Replica::FullSyncDflyFb(unique_ptr<base::IoBuf> io_buf, string eof_token) {
+error_code Replica::StartStableSyncFlow() {
+  DCHECK(!master_context_.master_repl_id.empty() && !master_context_.dfly_session_id.empty());
+  ProactorBase* mythread = ProactorBase::me();
+  CHECK(mythread);
+
+  CHECK(sock_->IsOpen());
+  // sock_.reset(mythread->CreateSocket());
+  // RETURN_ON_ERR(sock_->Connect(master_context_.master_ep));
+  sync_fb_ = ::boost::fibers::fiber(&Replica::StableSyncDflyFb, this);
+
+  return std::error_code{};
+}
+
+void Replica::FullSyncDflyFb(SyncBlock* sb, string eof_token) {
+  DCHECK(leftover_buf_);
   SocketSource ss{sock_.get()};
-  io::PrefixSource ps{io_buf->InputBuffer(), &ss};
+  io::PrefixSource ps{leftover_buf_->InputBuffer(), &ss};
 
   RdbLoader loader(NULL);
+  loader.fullsyncb = [this, sb, ran = false]() mutable {
+    if (ran) return;
+    {
+      std::unique_lock lk(sb->mu_);
+      sb->flows_left--;
+      ran = true;
+    }
+    sb->cv_.notify_all();
+  };
   loader.Load(&ps);
 
+  // Try finding eof token.
+  io::PrefixSource chained_tail{loader.Leftover(), &ps};
   if (!eof_token.empty()) {
-    unique_ptr<uint8_t[]> buf(new uint8_t[eof_token.size()]);
-    // pass leftover data from the loader.
-    io::PrefixSource chained(loader.Leftover(), &ps);
-    VLOG(1) << "Before reading from chained stream";
-    io::Result<size_t> eof_res = chained.Read(io::MutableBytes{buf.get(), eof_token.size()});
-    if (!eof_res || *eof_res != eof_token.size()) {
+    unique_ptr<uint8_t[]> buf{new uint8_t[eof_token.size()]};
+
+    io::Result<size_t> res =
+        chained_tail.ReadAtLeast(io::MutableBytes{buf.get(), eof_token.size()}, eof_token.size());
```

Comment:
You had a bug here: We need to use `ReadAtLeast` because the source might return less than n in a single call.

Reply:
Read can return less than buf.size() in a single call (see the PrefixSource read, it returns only the prefix even if its smaller than buf.size())